tests/gacha_tests.cpp

```diff
@@ -0,0 +1,162 @@
+   subopen({N(gacha3), N(conr2d)}, EA("5000.00 CRD@conr2d"));
+   open({N(gacha3), N(conr2d)}, gd, EA("5000.00 CRD@conr2d"), expiration, 1);
+   produce_blocks(1);
+
+   BOOST_REQUIRE_EQUAL(success(),
+         issue(N(eun2ce), {N(gacha3), N(conr2d)}, dseedhash)
+   );
+
+} FC_LOG_AND_RETHROW()
+
+BOOST_FIXTURE_TEST_CASE(setoseed_tests, gxc_gacha_tester) try {
+
+   vector<grade> gd = {{.reward = asset::from_string("5000.00 ENC"), .score = 1, .limit = 0}};
+   string expiration = "2022-01-01T00:00:10";
+   string dseedhash = "4855c552bbdd69cef5faee9bcc16a2f6a6b9d5f04560fa6a4045a5195c1f4f75";
+   string oseed = "4c519413ac98e5ead1c3b412e5d053ba0d57245e7689e6fcbe6dd9b81aa88dd7";
+
+   subopen({N(gachatest), N(eun2ce)}, EA("5000.00 ENC@eun2ce"));
+   open({N(gachatest), N(eun2ce)}, gd, EA("5000.00 ENC@eun2ce"), expiration, 1);
+   issue(N(conr2d), {N(gachatest), N(eun2ce)}, dseedhash, 1);
+   produce_blocks(1);
+
+   BOOST_REQUIRE_EQUAL(success(),
+         setoseed(1, oseed, N(conr2d))
+   );
+   produce_blocks(1);
+
+   string scmstr = "{\"scheme_name\": \"gachatest\", \"grades\": [{\"reward\": \"5000.00 ENC\", \"score\": 1, \"limit\": 0}], \"budget\": \"5000.00 ENC@eun2ce\", \"expiration\": \"2022-01-01T00:00:10\", \"precision\": 1, \"deadline_sec\": 604800, \"out\": \"0.00 ENC\", \"out_count\": [0], \"issued\": 1, \"unresolved\": 1}";
+   string gchstr = "{\"id\": 1, \"owner\": \"conr2d\", \"scheme\": {\"name\": \"gachatest\", \"contract\": \"eun2ce\"}, \"dseedhash\": \"4855c552bbdd69cef5faee9bcc16a2f6a6b9d5f04560fa6a4045a5195c1f4f75\", \"oseed\": \"4c519413ac98e5ead1c3b412e5d053ba0d57245e7689e6fcbe6dd9b81aa88dd7\", \"deadline\": \"2020-01-08T00:00:11\"}";
+
+   BOOST_REQUIRE_EQUAL(scmstr,
+         std::to_string(get_table_row(gacha_account_name, N(eun2ce), N(scheme), N(gachatest)))
+   );
+
+   BOOST_REQUIRE_EQUAL(gchstr,
+         std::to_string(get_table_row(gacha_account_name, gacha_account_name, N(gacha), 1, "_gacha"))
+   );
+
+   oseed = "4c519413ac98e5ead1c3b412e5d053ba0d57245e7689e6fcbe6dd9b81aa88dd6";
+   BOOST_REQUIRE_EQUAL(wasm_assert_msg("oseed is already set"),
+         setoseed(1, oseed, N(conr2d))
+   );
+
+} FC_LOG_AND_RETHROW()
+
+BOOST_FIXTURE_TEST_CASE(setdseed_tests, gxc_gacha_tester) try {
+
+   vector<grade> gd = {{.reward = asset::from_string("5000.00 ENC"), .score = 1, .limit = 0}};
+   string expiration = "2022-01-01T00:00:10";
+   string dseedhash = "4855c552bbdd69cef5faee9bcc16a2f6a6b9d5f04560fa6a4045a5195c1f4f75";
+   string oseed = "1ba0668f40b6fdd0a2b553a8e5dfe57ee977506532ba437f31f1b72abc199dbe";
+
+   subopen({N(gachatest), N(eun2ce)}, EA("5000.00 ENC@eun2ce"));
+   open({N(gachatest), N(eun2ce)}, gd, EA("5000.00 ENC@eun2ce"), expiration, 1);
+   issue(N(conr2d), {N(gachatest), N(eun2ce)}, dseedhash, 1);
+   setoseed(1, oseed, N(conr2d));
+   produce_blocks(1);
+
+   BOOST_REQUIRE_EQUAL(success(),
+      setdseed(1, "4c519413ac98e5ead1c3b412e5d053ba0d57245e7689e6fcbe6dd9b81aa88dd7")
```

Comment:
Check whether not only transaction is pushed, but reward is paid correctly (or not paid according to score). You can compute score in advance, because random number is generated deterministically.